src/datatrove/pipeline/stats/base.py

```diff
@@ -0,0 +1,112 @@
+import heapq
+import json
+from abc import abstractmethod
+from collections import defaultdict
+from typing import get_args
+
+from loguru import logger
+
+from datatrove.data import Document, DocumentsPipeline
+from datatrove.io import DataFolderLike, get_datafolder
+from datatrove.pipeline.base import PipelineStep
+from datatrove.pipeline.stats.config import DEFAULT_TOP_K_CONFIG, GROUP, STAT_TYPE, TopKConfig
+from datatrove.utils.stats import MetricStatsDict
+
+
+class BaseStats(PipelineStep):
+    """
+    Datatrove block for computing statistics of dataset.
+    Each stat is of type MetricStatsDict saved in output_folder/{group}/{stat_name}/{rank:05d}.json
+    Args:
+        output_folder: The folder where the statistics will be saved.
+        groups_to_compute: The groups of statistics to compute.
+        histogram_round_digits: The number of digits to round the histogram values to.
+            This ensures reasonable number of bins.
+        top_k_config: The configuration for compressing the statistics.
+            Each group in top_k_groups will truncate the statistics to the top k keys.
+            This lowers memory usage and speeds up the merging in second-stage.
+    """
+
+    type = "📊 - STATS"
+    name = "👑 Summary stats"
+    _requires_dependencies = ["tldextract"]
+
+    def __init__(
+        self,
+        output_folder: DataFolderLike,
+        groups_to_compute: list[GROUP] | None = None,
+        histogram_round_digits: int = 3,
+        top_k_config: TopKConfig = DEFAULT_TOP_K_CONFIG,
+    ) -> None:
+        from tldextract import TLDExtract
+
+        super().__init__()
+        self.output_folder = get_datafolder(output_folder)
+        self.groups = groups_to_compute or list(get_args(GROUP))
+        self.histogram_round_digits = histogram_round_digits
+        self.top_k_cfg = top_k_config
+        self.tld_extractor = TLDExtract()
+
+    @abstractmethod
+    def extract_stats(self, doc: Document) -> dict[str, int | float]:
+        """
+        Abstract method for extracting stats from a document.
+        Args:
+            doc: The document to extract stats from.
+
+        Returns:
+            A dictionary of statistics, where the key is the stat name and the value is the stat value.
+        """
+        raise NotImplementedError()
+
+    def get_kv(self, doc: Document, value: STAT_TYPE, group_name: GROUP) -> tuple[str, STAT_TYPE]:
+        if group_name == "histogram":
+            # Use rounding to reduce then number of values for histogram
+            return str(round(value, self.histogram_round_digits)), 1
+        elif group_name == "summary":
+            return "summary", value
+        elif group_name == "fqdn":
+            return doc.metadata.setdefault("fqdn", self.tld_extractor.extract_str(doc.metadata["url"]).fqdn), value
```

Comment:
does this not rerun tld_extractor every time, including when fqdn is already there?